prep release: v1.30.0 (#3828)
apollographql/router: CHANGELOG.md

This project adheres to [Semantic Versioning v2.0.0](https://semver.org/spec/v2.0.0.html).

# [1.30.0] - 2023-09-14

## 🚀 Features

> ### Preliminary Rhai Support at the `router_service` ([Issue #2278](https://github.com/apollographql/router/issues/2278))

Author: It's subject to improve in what we believe are obvious ways that we just haven't done yet 😛.  Concretely, it means that we don't support the `body` property yet, only `context` and `headers`.   It didn't say _Preliminary_ before, but I learned about that limit and decided to add the word.  I'll remove it. It's fine without it.

```suggestion
### Rhai Support at the `router_service` ([Issue #2278](https://github.com/apollographql/router/issues/2278))
```

To resolve a security advisory (for which our usage was _not_ affected), we've replaced `atty` with `std`.  Instead, we now use equivalent functionality available in the Rust standard library, available since Rust v1.70.0.

* https://github.com/apollographql/router/security/dependabot/68
* https://doc.rust-lang.org/stable/std/io/trait.IsTerminal.html

By [@SimonSapin](https://github.com/SimonSapin) in https://github.com/apollographql/router/pull/3729

### Upgrade `webpki` and `rustls-webpki` ([PR #3728](https://github.com/apollographql/router/pull/3728))

These two dependency updates brings fixes for two separate security advisories:

* https://rustsec.org/advisories/RUSTSEC-2023-0052
* https://rustsec.org/advisories/RUSTSEC-2023-0053

Since Apollo Router does _not_ accept client certificates, it could only have been affected if a subgraph had provided a pathological TLS server certificate.

By [@SimonSapin](https://github.com/SimonSapin) in https://github.com/apollographql/router/pull/3728

## 📚 Documentation

### GraphOS authorization: Exemplify scope manipulation with Rhai at the router service level ([PR #3719](https://github.com/apollographql/router/pull/3719))

> New [Authorization documentation](https://www.apollographql.com/docs/router/configuration/authorization/#requiresscopes) shows the how to use Rhai script to extract scopes and prepare them in the correct way, for use with `@requiresScope`.  This becomes relevant since `@requiresScopes` expects scopes to come from the `scope` claim in the [OAuth2 access token format](https://datatracker.ietf.org/doc/html/rfc6749#section-3.3) while tokens may have scopes stored differently, e.g., as an array of strings, or even as different claims.  If you have further questions on the right choice for you, please open a GitHub Discussion that provides an example of what you need to achieve.

Reviewer: Agree - can be simplified a little although I do like the CTA. 

New [Authorization documentation](https://www.apollographql.com/docs/router/configuration/authorization/#requiresscopes) shows how to use Rhai script to extract scopes and prepare them for use with the `@requiresScopes` directive. This is particularly useful when scopes are stored in a format that is different from the [Oath2 access token format](https://datatracker.ietf.org/doc/html/rfc6749#section-3.3). 

If you have further questions on the right choice for you, please open a GitHub Discussion that provides an example of what you need to achieve.

Still too techincal @BrynCooke?

> New [Authorization documentation](https://www.apollographql.com/docs/router/configuration/authorization/#requiresscopes) shows the how to use Rhai script to extract scopes and prepare them in the correct way, for use with `@requiresScope`.  This becomes relevant since `@requiresScopes` expects scopes to come from the `scope` claim in the [OAuth2 access token format](https://datatracker.ietf.org/doc/html/rfc6749#section-3.3) while tokens may have scopes stored differently, e.g., as an array of strings, or even as different claims.  If you have further questions on the right choice for you, please open a GitHub Discussion that provides an example of what you need to achieve.

Author: I think that is mostly what this text aims to do, to be honest.  As new as a feature this is, I enjoy giving it some prominence either way while **not** going as far as giving the answer as to **how** to do it.

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/3773

### Metrics are no longer coerced incorrectly ([Issue #3687](https://github.com/apollographql/router/issues/3687))

Metric attributes are no longer incorrectly coerced to strings.  In addition, the logic around types which are accepted as metrics attributes has been simplified to avoid this in the future.  Going forward, if the wrong type is specified, values will be ignored and a log message will be emitted.

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/3724

### Optimizations applied to header-handling operations ([Issue #3068](https://github.com/apollographql/router/issues/3068))

We've taken advantage of an opportunity to reduce latency and overhead by reducing the allocations which take place when handling the list of headers passed to subgraph queries.

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/3721

### Avoid request overhead when telemetry is not enabled

> We have removed overhead that was being incurred on every request within the OpenTelemetry code, even when telemetry was not enabled or configured.  The introduction of a sampling filter now disables it entirely when no OpenTelemetry exporters are enabled, providing an improvement for basic setups where telemetry is not desired.

Reviewer 2: ```suggestion
The overhead of OpenTelemetry has been removed when no tracing exporters and are configured and managed federation is not in use.
```

Author: Countering with:

```suggestion
The overhead of OpenTelemetry has been removed when no tracing exporters are configured.
```

> This improvement also manifests when the sampling criteria is _not_ met since the sampling filter will only propagate sampled traces to the rest of the stack, further reducing overhead.

Reviewer 2: ```suggestion
This also improves performance when sampling criteria has _not_ been met by preventing unsampled sampled trace events from propagating to the rest of the opentelemetry stack.
```

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/2999

### Subgraph authentication: Apply signature after compression and APQ ([Issue #3608](https://github.com/apollographql/router/issues/3608))

> The router will now _sign_ subgraph requests _just before_ they are sent to the subgraph (i.e., a bit later than previously), following up on the functionality of [subgraph authentication](https://www.apollographql.com/docs/router/configuration/authn-subgraph) which was first introduced in v1.27.0.

Reviewer 2: ```suggestion
Previously, subgraph requests could be modified they had been sigv4 signing had taken place, resulting is a signature mismatch. The router will now sign subgraph requests _just before_ they are sent to the subgraph.
```

While the initial implementation did work in many cases, there were interactions with other subgraph features which might be enabled which required additional considerations, including interactions with:

  - Subgraph Automatic Persisted Queries (APQ)
  - Subgraph HTTP compression
  - Custom plugins that operate on the subgraph service (whether via Co-Processors, Rhai or a compiled Rust plugin)

> In most cases, the interactions between these features and the subgraph authentication feature were problematic and required disabling one or the other in order to generate a request that was correctly signed by the signature algorithm.  This should all be resolved.

Reviewer 2: ```suggestion

```

By [@o0Ignition0o](https://github.com/o0Ignition0o) in https://github.com/apollographql/router/pull/3735

### Handle multipart stream if the original stream was empty ([Issue #3293](https://github.com/apollographql/router/issues/3293))

> Multi-part response streams (which are used for [subscriptions](https://www.apollographql.com/docs/router/executing-operations/subscription-support/) and operations which include [`@defer` directive](https://www.apollographql.com/docs/router/executing-operations/defer-support/)) will now be handled correctly (and end correctly) when the response stream is empty.

Reviewer 2: ```suggestion
Multi-part response streams (which are used for [subscriptions](https://www.apollographql.com/docs/router/executing-operations/subscription-support/) and operations which include [`@defer` directive](https://www.apollographql.com/docs/router/executing-operations/defer-support/)) are now terminated correctly when the response stream is empty.
```

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/3748

> ### Subscriptions: Include `x-accel-buffering`` header on multipart responses ([Issue #3683](https://github.com/apollographql/router/issues/3683))

Reviewer 3: ```suggestion
### Subscriptions: Include `x-accel-buffering` header on multipart responses ([Issue #3683](https://github.com/apollographql/router/issues/3683))
```

> Setting the `x-accel-buffering` header to `no` for multipart responses allows certain proxies to configure themselves in a mode that is compatible with the buffering used by subscriptions.  This should improve Subscriptions' compatibility with existing infrastructure.

Reviewer 2: ```suggestion
Setting the `x-accel-buffering` header to `no` for multipart responses allows certain proxies to configure themselves in a mode that is compatible with the buffering used by subscriptions.  This improves Subscriptions' compatibility with existing infrastructure.
```

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/3749

## 🛠 Maintenance

### Our Rust Toolchain has been updated to v1.72.0 ([PR #3707](https://github.com/apollographql/router/pull/3707))

> Our Rust Toolchain has been updated to v1.72.0 after we've unblocked some downstream dependency juggling within our `router-bridge` dependency.  For the majority of our users (those who do not compile their own Router from source), this change will not have any impact.  If you are compiling the Router using the new Rust 1.72.0, it should now work.

Reviewer 2: ```suggestion
Our Rust Toolchain has been updated to v1.72.0.  For the majority of our users (those who do not compile their own Router from source), this change will not have any impact. If you are compiling the Router using the new Rust 1.72.0, it will now work.
```

By [@o0Ignition0o](https://github.com/o0Ignition0o) in https://github.com/apollographql/router/pull/3707

### `Cargo.toml` metadata cleanup ([PR #3746](https://github.com/apollographql/router/pull/3746))

The following changes are not material for most users of the Router (especially those who are not compiling Rust themselves):

* Removed unused `[patch]` entries in in our `Cargo.toml`.
* Removed prior exemptions for [`chrono`](https://crates.io/crates/chrono) security advisories which are now resolved.

> 

Author: Removed per other suggestion for the same.  No disagreement.

This project adheres to [Semantic Versioning v2.0.0](https://semver.org/spec/v2.0.0.html).

# [1.30.0] - 2023-09-14

## 🚀 Features

### Rhai Support at the `router_service` ([Issue #2278](https://github.com/apollographql/router/issues/2278))

> It is now possible to interact with requests and responses at the `router_service` level [using Rhai-based customizations](https://www.apollographql.com/docs/router/customizations/rhai/). The functionality is very similar to that provided for interacting with existing services, for example `supergraph_service`. For instance, you may "map" requests and responses as follows:

Reviewer 2: ```suggestion
It is now possible to intercept requests and responses at the `router_service` level [using Rhai-based customizations](https://www.apollographql.com/docs/router/customizations/rhai/). The functionality mirrors existing functionality router services, for example `supergraph_service`. For example, "maping" requests and responses as follows:
```

Author: I'll resolve this as I don't think _intercept_ is actually the right terminology to describe how these layers work.  The other changes you suggest 
are almost identical.  (Maybe regional? e.g., for instance vs for example 😄 )

```rust
fn router_service(service) {
    const request_callback = Fn("process_request");
    service.map_request(request_callback);
    const response_callback = Fn("process_response");
    service.map_response(response_callback);
}
```

> The main difference from [existing services](https://www.apollographql.com/docs/router/customizations/rhai/#router-request-lifecycle) is that the `router_service` allows operating on _HTTP bodies_ rather than the more structured representations available at later service layers, like the [supergraph service](https://www.apollographql.com/docs/router/customizations/rhai/#supergraphservice).

Reviewer 2: ```suggestion
Unlike [existing services](https://www.apollographql.com/docs/router/customizations/rhai/#router-request-lifecycle), the `router_service` allows operating on _HTTP bodies_ rather than the more structured representations available at later service layers, like the [supergraph service](https://www.apollographql.com/docs/router/customizations/rhai/#supergraphservice).
```

This means that `Request.body` and `Response.body` are both strings, which provides flexibility to make more substantial modifications to the body.  Users should choose the most convenient service for the types of changes they're needing to make, since operating on the strings will require more work than operating on properties of the structured objects in the supergraph service.

As another example, this logs the bodies of requests:

```rust
// Generate a log for each request at this stage
fn process_request(request) {
    print(`body: ${request.body}`);
}

// Generate a log for each response at this stage
fn process_response(response) {
    print(`body: ${response.body}`);
}
```

By [@garypen](https://github.com/garypen) in https://github.com/apollographql/router/pull/3234

## 🐛 Fixes

### Small performance improvements to telemetry ([PR #3656](https://github.com/apollographql/router/pull/3656))

> We applied some small performance improvements to the `SpanMetricsExporter` (which is used to report span timings), some of which apply in cases where telemetry is disabled and could be apparent to most users.

Reviewer 2: ```suggestion
Small performance improvements to the `SpanMetricsExporter` (which is used to report span timings).
```

Author: I'm going to keep what I had on this one because I feel the resulting sentence is just too short / borderline incomplete, but I see where you were going.  I did want to de-emphasise the Rust aspect of it, but also emphasize the improvements for everyone.

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/3656

### Handle interfaces in fragment spreads when `__typename` is omitted ([Issue #2587](https://github.com/apollographql/router/issues/2587))

> We now check the parent type when using an inline-fragment, rather than relying on the expectation that `__typename` will be present, since that is not a guarantee.  For cases where `__typename` was being omitted, this corrects cases where a portion of the selection set was silently dropped and not returned.

Reviewer 2: ```suggestion
Parent type when using an inline-fragment is verified, rather than relying on the expectation that `__typename` will be present.  For cases where `__typename` was being omitted, this corrects responses where a portion of the selection set was silently dropped and not returned.
```

By [@o0Ignition0o](https://github.com/o0Ignition0o) and [@geal](https://github.com/geal) in https://github.com/apollographql/router/pull/3718

### Deduplication is, again, enabled by default as documented ([PR #3773](https://github.com/apollographql/router/pull/3773))

[Subscription deduplication](https://www.apollographql.com/docs/router/executing-operations/subscription-support#subscription-deduplication) is again enabled by default as it was intended to be.  This important performance feature for subscriptions at scale was inadvertently disabled in v1.25.0 due to a bug.

Should you still wish to disable it, you can still explicitly [set `enable_deduplication` to `false` in your configuration](https://www.apollographql.com/docs/router/executing-operations/subscription-support/#disabling-deduplication).

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/3773

### Metrics are no longer coerced incorrectly ([Issue #3687](https://github.com/apollographql/router/issues/3687))

> Metric attributes are no longer incorrectly coerced to strings.  In addition, the logic around types which are accepted as metrics attributes has been simplified to avoid this in the future.  Going forward, if the wrong type is specified, values will be ignored and a log message will be emitted.

Reviewer 2: ```suggestion
Metric attributes are no longer incorrectly coerced to strings.  In addition, the logic around types which are accepted as metrics attributes has been simplified to avoid this in the future.  Going forward, if the wrong type is specified, values will be ignored and a log message (at debug level) will be emitted.
```

By [@BrynCooke](https://github.com/BrynCooke) in https://github.com/apollographql/router/pull/3724

### Optimizations applied to header-handling operations ([Issue #3068](https://github.com/apollographql/router/issues/3068))

> We've taken advantage of an opportunity to reduce latency and overhead by reducing the allocations which take place when handling the list of headers passed to subgraph queries.

Reviewer 2: ```suggestion
Latency and overhead of passing headers to subgraph queries has been reduced.
```

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/3748

### Subscriptions: Include `x-accel-buffering`` header on multipart responses ([Issue #3683](https://github.com/apollographql/router/issues/3683))

Setting the `x-accel-buffering` header to `no` for multipart responses allows certain proxies to configure themselves in a mode that is compatible with the buffering used by subscriptions.  This should improve Subscriptions' compatibility with existing infrastructure.

By [@bnjjj](https://github.com/bnjjj) in https://github.com/apollographql/router/pull/3749

## 🛠 Maintenance

### Our Rust Toolchain has been updated to v1.72.0 ([PR #3707](https://github.com/apollographql/router/pull/3707))

Our Rust Toolchain has been updated to v1.72.0 after we've unblocked some downstream dependency juggling within our `router-bridge` dependency.  For the majority of our users (those who do not compile their own Router from source), this change will not have any impact.  If you are compiling the Router using the new Rust 1.72.0, it should now work.

By [@o0Ignition0o](https://github.com/o0Ignition0o) in https://github.com/apollographql/router/pull/3707

> ### `Cargo.toml` metadata cleanup ([PR #3746](https://github.com/apollographql/router/pull/3746))
> 
> The following changes are not material for most users of the Router (especially those who are not compiling Rust themselves):
> 
> * Removed unused `[patch]` entries in in our `Cargo.toml`.
> * Removed prior exemptions for [`chrono`](https://crates.io/crates/chrono) security advisories which are now resolved.
> 
> By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/3746

Reviewer 4: ```suggestion
```

let's remove that one, it's not really actionable for users

By [@SimonSapin](https://github.com/SimonSapin) in https://github.com/apollographql/router/pull/3728

## 📚 Documentation

### GraphOS authorization: Exemplify scope manipulation with Rhai at the router service level ([PR #3719](https://github.com/apollographql/router/pull/3719))

New [Authorization documentation](https://www.apollographql.com/docs/router/configuration/authorization/#requiresscopes) shows the how to use Rhai script to extract scopes and prepare them in the correct way, for use with `@requiresScope`.  This becomes relevant since `@requiresScopes` expects scopes to come from the `scope` claim in the [OAuth2 access token format](https://datatracker.ietf.org/doc/html/rfc6749#section-3.3) while tokens may have scopes stored differently, e.g., as an array of strings, or even as different claims.  If you have further questions on the right choice for you, please open a GitHub Discussion that provides an example of what you need to achieve.

By [@Geal](https://github.com/Geal) in https://github.com/apollographql/router/pull/3719

> ### Anchor links in CORS documentation have been fixed
> 
> This documentation change fixes an incorrect anchor link in the [CORS documentation](https://www.apollographql.com/docs/router/configuration/cors/) and removes links to authorization docs which have not yet been released.
> 
> By [@Meschreiber](https://github.com/Meschreiber) in https://github.com/apollographql/router/pull/3711

Reviewer 4: ```suggestion
```
This one should probably be removed too
